<a href="https://colab.research.google.com/github/vasanthireddyaraveeti-eng/demo/blob/main/salarypredection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df=pd.read_csv("/content/archive (3).zip")
df.head()

,Age,Gender,Education Level,Job Title,Years of Experience,Salary
0,32.0,Male,Bachelor's,Software Engineer,5.0,90000.0
1,28.0,Female,Master's,Data Analyst,3.0,65000.0
2,45.0,Male,PhD,Senior Manager,15.0,150000.0
3,36.0,Female,Bachelor's,Sales Associate,7.0,60000.0
4,52.0,Male,Master's,Director,20.0,200000.0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375 entries, 0 to 374
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  373 non-null    float64
 1   Gender               373 non-null    object 
 2   Education Level      373 non-null    object 
 3   Job Title            373 non-null    object 
 4   Years of Experience  373 non-null    float64
 5   Salary               373 non-null    float64
dtypes: float64(3), object(3)
memory usage: 17.7+ KB


In [ ]:
print(df.shape)

(375, 6)


In [ ]:
df.describe()

,Age,Years of Experience,Salary
count,373.000000,373.000000,373.000000
mean,37.431635,10.030831,100577.345845
std,7.069073,6.557007,48240.013482
min,23.000000,0.000000,350.000000
25%,31.000000,4.000000,55000.000000
50%,36.000000,9.000000,95000.000000
75%,44.000000,15.000000,140000.000000
max,53.000000,25.000000,250000.000000


In [ ]:
df.isna().sum()

,0
Age,2
Gender,2
Education Level,2
Job Title,2
Years of Experience,2
Salary,2


In [ ]:
med = df["Age"].median()
df["Age"] = df["Age"].fillna(med)
df.head(25)

,Age,Gender,Education Level,Job Title,Years of Experience,Salary
0,32.0,Male,Bachelor's,Software Engineer,5.0,90000.0
1,28.0,Female,Master's,Data Analyst,3.0,65000.0
2,45.0,Male,PhD,Senior Manager,15.0,150000.0
3,36.0,Female,Bachelor's,Sales Associate,7.0,60000.0
4,52.0,Male,Master's,Director,20.0,200000.0
5,29.0,Male,Bachelor's,Marketing Analyst,2.0,55000.0
6,42.0,Female,Master's,Product Manager,12.0,120000.0
7,31.0,Male,Bachelor's,Sales Manager,4.0,80000.0
8,26.0,Female,Bachelor's,Marketing Coordinator,1.0,45000.0
9,38.0,Male,PhD,Senior Scientist,10.0,110000.0


In [ ]:
df.isna().sum()

,0
Age,0
Gender,2
Education Level,2
Job Title,2
Years of Experience,2
Salary,2


In [ ]:
df =df.dropna()

In [ ]:
df.isna().sum()

,0
Age,0
Gender,2
Education Level,2
Job Title,2
Years of Experience,2
Salary,2


In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df['Job Title'] = le.fit_transform(df['Job Title'])
print(df.head())

    Age  Gender Education Level  Job Title  Years of Experience    Salary
0  32.0    Male      Bachelor's        159                  5.0   90000.0
1  28.0  Female        Master's         17                  3.0   65000.0
2  45.0    Male             PhD        130                 15.0  150000.0
3  36.0  Female      Bachelor's        101                  7.0   60000.0
4  52.0    Male        Master's         22                 20.0  200000.0


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [ ]:
x = df[['Job Title', 'Years of Experience']]
y = df.Salary

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [ ]:
mlr = LinearRegression()
mlr.fit(x_train,y_train)

LinearRegression()

In [ ]:
y_pred=mlr.predict(x_test)
print(y_pred)

[171120.32707437  99498.12711426 140449.88734437  74153.87288465
 148684.52164895 155788.09252968 181132.34675928 102568.15640756
  51065.77486213  94710.54324427 122532.36497373 175105.97913936
  43585.18284012  93794.92047258 129043.47406102  51065.77486213
  69635.58982986  41218.32098664  82011.48604794  86475.90893969
 103052.89787492  94118.08145082 140988.48897478  64626.5946671
  74153.87288465  57738.46443853 155411.07138839  43369.74218796
 122101.48366941 157613.36743244  55967.04969881 129151.1943871
  52089.1179599  134800.54086573 100898.49135331  33899.30945375
 175752.30109585  43100.44137276 156703.71530136  56822.84166685
  44123.78447052  95249.14487467 197003.18293439 157559.5072694
  41538.49664458 143735.35728984 106655.55815802  49886.82191585
  45093.26740525  41269.19582938 157290.2064542   87337.67154834
 123555.7080715  116559.85751685  95033.70422251 114890.1924626
 203897.28380356  80072.52017849 171120.32707437  76092.8387541
 116829.15833206 136362.485593

In [ ]:
print("R2 Score:", r2_score(y_test,y_pred))

R2 Score: 0.8974889742185352


In [ ]:
xnew =np.array([[159, 2]])
prediction =mlr.predict(xnew)
print("Prediction:", prediction)

Prediction: [49294.3601224]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## Project Report: Salary Prediction using Linear Regression

### 1. Project Title
Salary Prediction Model using Machine Learning

### 2. Objective
The primary objective of this project is to develop a machine learning model capable of predicting an individual's salary based on key professional attributes such as their Job Title and Years of Experience. This model aims to provide insights into salary determinants and offer a predictive tool for estimating potential earnings.

### 3. Dataset Description
The project utilizes a dataset loaded from `/content/archive (3).zip`. The dataset contains information about individuals, including:
*   **Age**: The age of the individual (float).
*   **Gender**: The gender of the individual (object).
*   **Education Level**: The educational attainment of the individual (object).
*   **Job Title**: The professional role of the individual (object).
*   **Years of Experience**: The total years of professional experience (float).
*   **Salary**: The salary of the individual (float), which is the target variable for prediction.

The dataset initially contains 375 entries and 6 columns. After preprocessing, 373 valid entries were used for model training.

### 4. Data Preprocessing Steps Performed
The following steps were taken to clean and prepare the data:

*   **Loading Data**: The dataset was loaded into a pandas DataFrame from a CSV file within a zip archive.
*   **Missing Value Identification**: `df.isna().sum()` was used to identify columns with missing values. It was found that 'Age', 'Gender', 'Education Level', 'Job Title', 'Years of Experience', and 'Salary' each had 2 missing values.
*   **Handling Missing Values (Age)**: The missing values in the 'Age' column were imputed using the median age of the dataset (36.0).
*   **Handling Missing Values (Remaining Columns)**: All rows containing any remaining `NaN` values across other columns were dropped using `df.dropna()`. This ensured that the model was trained on complete records.
*   **Feature Encoding ('Job Title')**: The 'Job Title' column, being a categorical feature, was transformed into numerical format using `LabelEncoder` from `sklearn.preprocessing`. Each unique job title was assigned a unique integer value.

### 5. Feature Selection Process
For this linear regression model, the following features were selected from the preprocessed dataset as independent variables (X):

*   `Job Title` (after Label Encoding)
*   `Years of Experience`

The `Salary` column was chosen as the dependent variable (Y), which the model aims to predict.

### 6. Machine Learning Algorithms Used

*   **Linear Regression**: A simple yet effective supervised machine learning algorithm, `LinearRegression` from `sklearn.linear_model`, was chosen for this predictive task. It models the relationship between a dependent variable and one or more independent variables by fitting a linear equation to the observed data.

### 7. Model Training and Testing Procedure

1.  **Data Splitting**: The preprocessed data was split into training and testing sets using `train_test_split` from `sklearn.model_selection`. A test size of 20% was used, with `random_state=42` for reproducibility.
    *   `x_train`, `y_train`: Used to train the model (298 samples).
    *   `x_test`, `y_test`: Used to evaluate the model's performance on unseen data (75 samples).
2.  **Model Initialization**: An instance of `LinearRegression` was created.
3.  **Model Training**: The linear regression model was trained using the `x_train` and `y_train` datasets (`mlr.fit(x_train, y_train)`).
4.  **Prediction**: After training, the model made predictions on the test set (`y_pred = mlr.predict(x_test)`).

### 8. Accuracy of Each Model

*   **Linear Regression**: The model achieved an R2 Score of **0.8975** (approximately 89.75%).

### 9. Best Performing Model and Justification

Since only one model, Linear Regression, was employed in this project, it is considered the best performing model by default. The R2 score of approximately 0.8975 indicates that the model explains about 89.75% of the variance in salary, which is a strong performance for a simple linear model. This suggests that 'Job Title' and 'Years of Experience' are significant predictors of salary within this dataset.

### 10. Sample Prediction Explanation

A sample prediction was performed for a new individual with specific characteristics:

*   `Job Title` (encoded): 159
*   `Years of Experience`: 2

The model predicted a salary of **49294.36** for this individual. This demonstrates how the trained model can be used to estimate salaries for new, unseen data points based on their job title (encoded) and years of experience.

### 11. Conclusion

This project successfully developed a linear regression model to predict salaries based on job title and years of experience. The model achieved a high R2 score, indicating its effectiveness in capturing the relationship between the chosen features and salary. The data preprocessing steps, including handling missing values and encoding categorical features, were crucial for preparing the data for model training.

### 12. Future Scope

To further enhance this project, the following areas could be explored:

*   **Feature Engineering**: Explore creating new features, such as combining 'Education Level' with 'Years of Experience' or creating interaction terms.
*   **Advanced Encoding**: Investigate more sophisticated encoding techniques for 'Job Title' and 'Education Level' (e.g., One-Hot Encoding, Target Encoding) to potentially improve model performance.
*   **Other Machine Learning Algorithms**: Experiment with more advanced regression algorithms such as Random Forest Regressor, Gradient Boosting Regressor, or Neural Networks to see if they can achieve higher accuracy.
*   **Hyperparameter Tuning**: Optimize the hyperparameters of the chosen model(s) to further improve performance.
*   **More Data**: If available, incorporating a larger and more diverse dataset could lead to a more robust and generalizable model.
*   **Error Analysis**: Analyze the residuals to understand where the model performs poorly and identify potential biases or areas for improvement.